pkg/bloomtokenizer/tokenizer.go

```diff
@@ -0,0 +1,175 @@
+package bloomtokenizer
+
+import (
+	"encoding/binary"
+	"unicode/utf8"
+
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+type Token struct {
+	Key   []byte
+	Value string
+}
+
+type Tokenizer interface {
+	Tokens(line string) []Token
+	getSkip() int
+	getMin() int
+	getMax() int
+}
+
+const TokenBufferSize = 4096
+
+type ngramTokenizer struct {
+	// [min,max) exclusivity
+	min, max, skip      int
+	buffers             [][]rune // circular buffers used for ngram generation
+	runeBuffer          []byte   // buffer used for token generation
+	tokenBuffer         []Token  // buffer used for holding tokens that is returned
+	internalTokenBuffer []Token  // circular buffer for tokens
+}
+
+func newNGramTokenizer(min, max, skip int) *ngramTokenizer {
+	capacity := max - min
+	t := &ngramTokenizer{
+		min:     min,
+		max:     max,
+		skip:    skip,
+		buffers: make([][]rune, capacity),
+	}
+	for i := t.min; i < t.max; i++ {
+		t.buffers[i-t.min] = make([]rune, i)
+	}
+	t.runeBuffer = make([]byte, 0, max*4)
+	t.tokenBuffer = make([]Token, 0, TokenBufferSize)
+	t.internalTokenBuffer = make([]Token, 0, TokenBufferSize)
+	for i := 0; i < cap(t.internalTokenBuffer); i++ {
+		tok := Token{}
+		tok.Key = make([]byte, 0, 132)
+		t.internalTokenBuffer = append(t.internalTokenBuffer, tok)
+	}
+
+	return t
+}
+
+func (t *ngramTokenizer) getSkip() int {
+	return t.skip
+}
+
+func (t *ngramTokenizer) getMin() int {
+	return t.min
+}
+
+func (t *ngramTokenizer) getMax() int {
+	return t.max
+}
+
+func (t *ngramTokenizer) Tokens(line string) []Token {
+	t.tokenBuffer = t.tokenBuffer[:0] // Reset the result slice
+	var i int                         // rune index (not position that is measured in the range loop)
+	numToks := 0
+	for _, r := range line {
+
+		// j is the index of the buffer to use
+		for j := 0; j < (t.max - t.min); j++ {
+			// n is the length of the ngram
+			n := j + t.min
+			// pos is the position in the buffer to overwrite
+			pos := i % n
+			t.buffers[j][pos] = r
+
+			if i >= n-1 && (i+1-n)%(t.skip+1) == 0 {
+				t.runeBuffer = reassemble(t.buffers[j], (i+1)%n, t.runeBuffer)
+				if numToks >= cap(t.internalTokenBuffer) || numToks == len(t.internalTokenBuffer) {
+					tok := Token{}
+					tok.Key = make([]byte, 0, 132) // Using a 4 byte token and a chunk identifier, it's really 31 bytes. Adding in for special chars and the like here
+					t.internalTokenBuffer = append(t.internalTokenBuffer, tok)
+				}
+				t.internalTokenBuffer[numToks].Key = t.internalTokenBuffer[numToks].Key[:0]
+				t.internalTokenBuffer[numToks].Key = append(t.internalTokenBuffer[numToks].Key, t.runeBuffer...)
+				t.internalTokenBuffer[numToks].Value = string(t.internalTokenBuffer[numToks].Key)
+				numToks++
+			}
+		}
+		i++
+	}
+	t.tokenBuffer = append(t.tokenBuffer, t.internalTokenBuffer[:numToks]...)
```

Comment:
What's the advantage of having `tokenBuffer` vs `internalTokenBuffer`? Because we reset the tokenbuffer on each `Tokens()` call, there's no protection against mutability issues. Using one slice would be strictly more efficient here afaict.